agent-connection/src/main/kotlin/com/luxoft/blockchainlab/corda/hyperledger/indy/PythonRefAgentConnection.kt

```diff
@@ -0,0 +1,233 @@
+package com.luxoft.blockchainlab.corda.hyperledger.indy
+
+import com.fasterxml.jackson.annotation.JsonProperty
+import com.fasterxml.jackson.databind.node.ObjectNode
+import com.luxoft.blockchainlab.hyperledger.indy.*
+import com.luxoft.blockchainlab.hyperledger.indy.utils.SerializationUtils
+import mu.KotlinLogging
+import org.java_websocket.client.WebSocketClient
+import org.java_websocket.handshake.ServerHandshake
+import java.lang.Thread.sleep
+import java.net.URI
+import java.time.Instant
+import java.util.*
+import kotlin.RuntimeException
+
+
+class PythonRefAgentConnection : AgentConnection {
+    override fun disconnect() {
+        if (getConnectionStatus() == AgentConnectionStatus.AGENT_CONNECTION_CONNECTED) {
+            webSocket.closeBlocking()
+            connectionStatus = AgentConnectionStatus.AGENT_CONNECTION_DISCONNECTED
+        }
+    }
+
+    override fun connect(url: String, login: String, password: String) {
+        disconnect()
+        webSocket = AgentWebSocketClient(URI(url))
+        webSocket.apply {
+            connectBlocking()
+            sendJson(StateRequest())
+            var stateResponse = waitForMessageOfType<State>(MESSAGE_TYPES.STATE_RESPONSE)
+            if(!checkState(stateResponse, login)) {
+                sendJson(WalletConnect(login, password, id = Instant.now().toEpochMilli().toString()))
+                stateResponse = waitForMessageOfType<State>(MESSAGE_TYPES.STATE_RESPONSE)
+            }
+            if(!checkState(stateResponse, login))
+                throw AgentConnectionException("Error connecting to $url")
+        }
+    }
+
+    private var connectionStatus : AgentConnectionStatus = AgentConnectionStatus.AGENT_CONNECTION_DISCONNECTED
+
+    override fun getConnectionStatus(): AgentConnectionStatus = connectionStatus
+
+    private var counterParty: IndyParty? = null
+
+    private lateinit var webSocket : AgentWebSocketClient
+
+    private fun checkState(stateMessage: State?, userName: String) : Boolean {
+        return if(stateMessage != null &&
+                stateMessage.content?.get("initialized") == true &&
+                stateMessage.content?.get("agent_name") == userName ) {
+            connectionStatus = AgentConnectionStatus.AGENT_CONNECTION_CONNECTED
+            true
+        } else {
+            connectionStatus = AgentConnectionStatus.AGENT_CONNECTION_DISCONNECTED
```

Comment:
"Connection" here is an abstraction. It just means that we've found the target agent and passed through some handshake with it. "Real physical connection" is re-established each time when it drops automatically.

I don't know how the connection behaves when a fatal error occurs. Maybe we should make the connection persistent. Or maybe we should close it after some inactivity period. Or maybe both. We need more test.

Reply:
Fault handling is not designed at the moment, some cases(like reconnection) will be handled by web socket client, but other require deep use case analysis.

Reply:
Let's put TODO here regarding discussion above and add issue/research task

Reply:
Pythonic indy-agent has 2 states: 'initialized' and 'not initialized' (logged in/out). Agent sends it's state within a STATE message, in response to the STATE_REQUEST. If the indy-agent has 'initialized' state, the client _must not_ send the CONNECT message (along with the login/password) because this causes an error. Otherwise, the agent awaits for a CONNECT message, and replies with STATE. The client must not send the CONNECT prior to checking the state. If the state is 'not initialized', the client sends a CONNECT and checks with the STATE received in reply. This function only checks the state (logged in/out). Disconnect is not yet implemented in the agent. The only way to reset the agent state is to restart the agent. The websocket connection is maintained automatically, it's re-established by demand if broken.